src/microcosm.js

```diff
@@ -193,37 +193,61 @@ class Microcosm extends Emitter implements Domain {
     return this.initial == null ? {} : this.initial
   }
 
-  recall(action: ?Action) {
+  recall(action: ?Action): Object {
     if (action && action.id in this.snapshots) {
-      return this.snapshots[action.id]
+      return this.snapshots[action.id].next
     }
 
-    return this.initial
+    return this.getInitialState()
+  }
+
+  /**
+   * Get the state prior to a given action, but include any upstream
+   * updates from parent Microcosms.
+   */
+  rebase(action: Action): Object {
+    let state = this.recall(action.parent)
+
+    if (this.parent) {
+      state = merge(state, this.parent.recall(action))
+    }
+
+    return state
   }
 
   /**
    * Create the initial state snapshot for an action. This is important so
    * that, when rolling back to this action, it always has a state value.
    */
-  createSnapshot(action: Action) {
-    this.snapshots[action.id] = this.recall(action.parent)
+  createSnapshot(action: Action): Snapshot {
+    let snapshot: Snapshot = {
+      last: this.state,
+      next: this.state,
+      status: 'inactive',
+      payload: undefined
+    }
+
+    this.snapshots[action.id] = snapshot
+
+    return snapshot
   }
 
   /**
    * Update the state snapshot for a given action
    */
   updateSnapshot(action: Action) {
-    let next = this.recall(action.parent)
-
-    if (this.parent) {
-      next = merge(next, this.parent.recall(action))
-    }
+    let snap = this.snapshots[action.id]
+    let last = this.rebase(action)
 
     if (!action.disabled) {
-      next = this.domains.dispatch(next, action)
+      snap.next = this.domains.dispatch(action, last, snap)
+    } else {
+      snap.next = last
     }
 
-    this.state = this.snapshots[action.id] = next
+    this.snapshots[action.id] = merge(snap, { last, status: action.status, payload: action.payload })
```

Comment:
Since we share data, we can't use nested mutation. The overhead here appears to be negligible.